### ISSUE TITLE
Применение метода оптимизации Adam

### ISSUE THREAD

**Issue by igsha** [helenross636/magic_season#10]
SGD показывает довольно медленный рост показателей. Нужно попробовать что-то более современное, например, Adam.
Привести обучение моделей с использованием Adam. Подогнать количество эпох под каждую модель так, чтобы были примерно одинаковые результаты качества распознавания.

**Comment by helenross636** [comment]
наименьшая сеть обучилась примерно с той же точностью, а вот на 1,5к перегнала ее более чем на процент за 1 эпоху.
https://github.com/helenross636/magic_season/blob/main/ALL_NETS.ipynb

**Comment by igsha** [comment]
* Построить для каждой модели историю `history = model.fit(...)` для выявления переобучения и момента, когда обучение можно раньше остановить
    ```python
    plt.plot(history.history['loss'], label='loss')
    plt.plot(history.history['val_loss'], label='val_loss')
    plt.legend()
    plt.grid()
    ```
* Использовать параметр `validate_split=0.1` в аргументах метода `fit`

**Comment by helenross636** [comment]
Почему-то строятся пустые графики (смотреть модели 2-4). Может надо делать метку metrics=["loss"] в model.compile(...)?
https://github.com/helenross636/magic_season/blob/main/ALL_NETS_history.ipynb

**Comment by igsha** [comment]
* В первом случае Вы написали `history = model.compile(...)`, а не `history = model.fit(...)`
* Используйте `batch_size=128`, чтобы побыстрее обучение шло
* Сократите количество эпох для 30-ти нейронов (например, до 300). Есть подозрение, что так тоже хорошо получится
* Для модели из 30 нейронов не построены графики истории и нет валидационной выборки
* Введите дополнительный первый слой `InputLayer(shape=(784,))`. Не нужно передавать во все слои `input_dim=784`, так как у каждого слоя своя должна быть размерность, которая рассчитывается автоматически
* На графиках ничего нет, потому что используется только 1 эпоха у всех остальных сетей. История строится по эпохам.

**Comment by helenross636** [comment]
- Из-за того, что убрала `input_dim=784`, не могу теперь выводить информацию по модели через `print(model.summary())`.
- После сокращения эпох и увеличения бач сайза, модель обучилась за 5 с лишним минут при точности 96.43%.
https://github.com/helenross636/magic_season/blob/main/ALL_NETS_history.ipynb

**Comment by igsha** [comment]
> Из-за того, что убрала input_dim=784, не могу теперь выводить информацию по модели через print(model.summary()).

Какая ошибка-то? Насколько я вижу, Вы создали слой, но никуда его не добавили.

**Comment by helenross636** [comment]
Теперь добавила, но не уверена, что он все таки добавился, потому что в структуре его не видно
https://github.com/helenross636/cvclasses19/blob/master/%22ALL_NETS_history_ipynb%22.ipynb

**Comment by igsha** [comment]
Вы добавили код в какой-то левый репозиторий `cvclasses19`.

Вы добавили не тот класс. Нужен https://keras.io/api/layers/core_layers/input/. Посмотрите, как Вы делали 4-ю лабораторную работу https://github.com/cvlabmiet/machine-learning/blob/main/lab4.ipynb. Там применялся класс `tf.keras.Input`.

**Comment by helenross636** [comment]
В той лабораторной создавался новый класс модели. 
https://github.com/helenross636/magic_season/blob/main/ALL_NETS_v2_0.ipynb

**Comment by igsha** [comment]
Наступает сильное переобучение. Попробуйте для моделей использовать следующее количество эпох:
```python
epochs = {30: 50, 1500: 10, 5000: 10, 10000:10}
```
Нам нужно посмотреть, что происходит с остальными моделями.

**Comment by helenross636** [comment]
Готово
https://github.com/helenross636/magic_season/blob/main/ALL_NETS_v2_0.ipynb

**Comment by igsha** [comment]
Хорошо. Видно, что модели 1500, 5000, 10000 ведут себя абсолютно одинаково. Видимо, модель 30 недостаточно хороша. Мы видим её предел на 96%. Осталвяем пока так. Работаем дальше только с этими моделями.
